tests/equilibria/test_file.py

```diff
@@ -97,67 +112,118 @@ def read_flat_array(fortran_format, array_size):
+        # This demands stricter adherence to the G-EQDSK
+        # format than bluemira's main reader.
+        self.read_strict_geqdsk(fname)
+
+        # Write data read in from test file into a new JSON
+        # file, with the suffix "_temp"
+        jname = os.path.splitext(fname)[0] + ".json"
+        eqdsk.write(jname, format="json")
+        d3 = EQDSKInterface.from_file(jname).to_dict()
+
+        # Clean up temporary files
+        os.remove(fname)
+        os.remove(jname)
+
+        # Compare dictionaries to check data hasn't
+        # been changed.
+        assert compare_dicts(d1, d2, verbose=True)
+        assert compare_dicts(d1, d3, verbose=True)
+        assert compare_dicts(d2, d3, verbose=True)
+
+    def test_read_matches_values_in_file(self):
+        eq = EQDSKInterface.from_file(self.testfiles[0])
+
+        assert eq.nz == 151
+        assert eq.nx == 151
+        assert eq.xdim == pytest.approx(3.14981545)
+        assert eq.ncoil == 0
+        assert eq.xc.size == 0
+        assert eq.nbdry == 72
+        np.testing.assert_allclose(
+            eq.xbdry[:3], [0.399993127e01, 0.399150254e01, 0.396906908e01]
+        )
+        np.testing.assert_allclose(
+            eq.zbdry[-3:], [-0.507187454e00, -0.240712636e00, 0.263892047e-01]
+        )
+
+    def test_values_match_annotated_types(self):
+        eq = EQDSKInterface.from_file(self.testfiles[0])
+
+        mismatched = []
+        for key, value_type in EQDSKInterface.__annotations__.items():
+            value = getattr(eq, key)
+            try:
+                check_type(key, value, value_type)
+            except TypeError:
+                mismatched.append((key, type(value), value_type))
+        assert not mismatched
+
+    def test_write_then_read_in_json_format(self):
+        eq = EQDSKInterface.from_file(self.testfiles[0])
+
+        with mock.patch(OPEN, new_callable=mock.mock_open) as open_mock:
+            eq.write("some/path.json", format="json")
+        written = combine_text_mock_write_calls(open_mock)
+
+        with mock.patch(
+            OPEN, new_callable=mock.mock_open, read_data=written
+        ) as open_mock:
+            eq2 = EQDSKInterface.from_file("/some/path.json")
+
+        assert eq2.nz == 151
+        assert eq2.nbdry == 72
+
+    def test_derived_field_is_calculated_if_not_given(self):
+        data = copy.deepcopy(self.eudemo_sof_data)
+        for field in ["x", "z", "psinorm"]:
+            del data[field]
+
+        open_mock = mock.mock_open(read_data=json.dumps(data))
+        with mock.patch(OPEN, new=open_mock):
+            eqdsk = EQDSKInterface.from_file("/some/file.json")
+
+        np.testing.assert_allclose(eqdsk.x, self.eudemo_sof_data["x"])
+        np.testing.assert_allclose(eqdsk.z, self.eudemo_sof_data["z"])
+        # TODO(hsaunders1904): understand why this fails
+        # np.testing.assert_allclose(eqdsk.psinorm, self.eudemo_sof_data["psinorm"])
```

Comment:
I can't remember why, or indeed if there was a reason, but when looking through our JSON files, it seems that the psinorm vector doesn't end in 1.0 (i.e. endpoint is discarded). In `_derive_psinorm` you have `np.linspace(0, 1, len(fpol))`, which would end in 1.0. I vaguely remember that we used not to be able to find the flux surface at psi_norm=1.0 very accurately using normal flux surface finding approaches. We since added the `find_LCFS_separatrix` which does the job. It could be because of this that we used to discard the 1.0 point in EQSDSKs. This feels like the kind of hack I might have 5 years ago... :D

Reply:
This seems to be it (well remembered!). Calculating `psinorm` as `np.linspace(0, 1, len(fpol) + 1)[:-1]` makes the test pass. I've updated the test so that we're checking more than just the size